src/sancus_support/sm_support.h

```diff
@@ -566,6 +572,7 @@ always_inline sm_id sancus_get_caller_id(void)
 {
```

Comment:
sancus_get_caller_id should be replaced with either a compiler intrinsic or (prob easier/better) a hand-coded asm function in the sm_entry stub file to return the stored caller ID from the SSA frame right?

Reply:
Good point and this is an interesting one. We already store the caller id in the `sm_ssa_caller_id` field that is at ssa_bas + 4. However, this means that we need the SM name to get to this address. After looking at the other functions in sm_support.h, my first instinct was to add the below function:
```C
always_inline sm_id sancus_get_original_caller_id(struct SancusModule* sm)
{
    sm_id ret;
    asm("mov %1 r15\n\t"
        "mov r15 %0"
        : "=m"(ret)
        // secret_end - 2 is ssa_base_addr that points to ssa_base. 
        // We want sm_caller_id that lies at ssa_base + 4
        : "m"(*( (void **)(sm->secret_end - 2)) + 4)
        : "15"
    );
    return ret;
}
```

This would technically work with the downside that it is an additional function (and not backward compatible). However, this calculation would rely on the untrusted `SancusModule` struct and I am pretty sure that the attacker could modify that pointer temporarily to trick the SM. Interestingly, the sm_support.h function `sancus_verify` should have that same problem, correct?

I see two more options: 
1. Implement get_caller_id as a compiler define that takes the sancus name and inserts it into the `sm_sancus_##name##_caller_id` address.
2. Automatically generate a `sm_##name##_get_caller_id` with each SM_DEFINE call.

I'm not too sure which is the best option (or whether I missed one)

Reply:
yes this is an interesting problem, I should probably have clarified a bit better what I meant above:

* compiler intrinsic:  the function is defined but not implemented here, during the compiler pass (when the name of the enclave is known), the compiler intercepts all invocations and pastes in the required asm instruction (parameterized with the enclave name). this is essentially a big ugly hack, but the developer luckily doesn't see it :)
* prob easier/better: a hand-coded asm function in the sm_entry stub file. Just declare the C function in the header file, and provide a hand-written asm implementation somewhere in the sm_entry/exit files (or provide a new file as with sm_verify.S). Then you can use a generic symbol name __sm_ssa or so, which is rewritten by our custom linker to include the enclave name (such a symbol already exists I think)

hope this makes it clearer?

Reply:
for reference, for the compiler intrinsic, I found my old commit in an old downstream version of the Sancus compiler to do exactly this (even though I think the 2nd approach above might be somewhat neater, more maintainable): 

```diff
commit cc54aa451bd41c183218fc91211ae6d2d6e06efd
Author: Jo Van Bulck <jo.vanbulck@cs.kuleuven.be>
Date:   Thu May 26 14:56:31 2016 +0200

    Implemented sancus_get_caller_id as a compiler intrinsic.
    
    CallerID is stored atomically in SM-private TCS memory on SM entry
    to ensure it remains stable in a preemptive environment. Compiler-
    generated sancus_get_caller_id function transparently fetches the
    callerID from the current SM TCS.

diff --git a/src/llvm/SancusModuleCreator.cpp b/src/llvm/SancusModuleCreator.cpp
index 8b54108..c61c7f4 100644
--- a/src/llvm/SancusModuleCreator.cpp
+++ b/src/llvm/SancusModuleCreator.cpp
@@ -42,6 +42,7 @@ struct SancusModuleCreator : ModulePass
     Constant* getSymbolAddress(Module& m, StringRef name);
     std::string fixSymbolName(const std::string& name);
     CallSite handleSancusCall(CallSite cs);
+    CallSite handleSancusGetCallerId(CallSite cs);
 
     typedef std::map<const GlobalValue*, SancusModuleInfo> InfoMap;
     InfoMap modulesInfo;
@@ -166,6 +167,12 @@ bool SancusModuleCreator::handleFunction(Function& f)
                 handleSancusCall(cs).getInstruction();
             continue;
         }
+        else if (callee->getName() == "sancus_get_caller_id")
+        {
+            replacements[cs.getInstruction()] =
+                handleSancusGetCallerId(cs).getInstruction();
+            continue;
+        }
 
         Function* stub = getStub(f, *callee);
         cs.setCalledFunction(stub);
@@ -531,3 +538,24 @@ CallSite SancusModuleCreator::handleSancusCall(CallSite cs)
     auto asmCall = CallInst::Create(inlineAsm, args);
     return CallSite(asmCall);
 }
+
+CallSite SancusModuleCreator::handleSancusGetCallerId(CallSite cs)
+{
+    auto callerInfo = getSancusModuleInfo(cs.getCaller());
+
+    /* callerID is stored in SM Thread Control Structure, below TCS stack */
+    auto asmStr = Twine("mov &" + callerInfo.getStackPointerAddress() + ", $0\n\t"
+                        "sub #" + callerInfo.getStackSize() + "+2, $0\n\t"
+                        "mov @$0, $0"
+                        ).str();
+    auto constraintsStr = "=r";
+    auto asmFuncTy = FunctionType::get(wordTy, /*isVarArg=*/false);
+
+    // NOTE hasSideEffects has to be true because the inline assembly may be
+    // optimised away otherwise
+    auto inlineAsm = InlineAsm::get(asmFuncTy, asmStr, constraintsStr,
+                                    /*hasSideEffects=*/true);
+
+    auto asmCall = CallInst::Create(inlineAsm);
+    return CallSite(asmCall);
+}
diff --git a/src/llvm/SancusModuleInfo.cpp b/src/llvm/SancusModuleInfo.cpp
index 0ce4262..77f4177 100644
--- a/src/llvm/SancusModuleInfo.cpp
+++ b/src/llvm/SancusModuleInfo.cpp
@@ -25,6 +25,16 @@ std::string SancusModuleInfo::getExitName() const
     return "__sm_" + name + "_exit";
 }
 
+std::string SancusModuleInfo::getStackSize() const
+{
+    return "__sm_" + name + "_stack_size";
+}
+
+std::string SancusModuleInfo::getStackPointerAddress() const
+{
+    return "__sm_" + name + "_sp_addr";
+}
+
 std::string SancusModuleInfo::getIndexName(const std::string& entry) const
 {
     return "__sm_" + name + "_entry_" + entry + "_idx";
diff --git a/src/llvm/SancusModuleInfo.h b/src/llvm/SancusModuleInfo.h
index abc3c48..d1ec81b 100644
--- a/src/llvm/SancusModuleInfo.h
+++ b/src/llvm/SancusModuleInfo.h
@@ -14,6 +14,8 @@ struct SancusModuleInfo
     std::string getTableSection() const;
     std::string getEntryName() const;
     std::string getExitName() const;
+    std::string getStackSize() const;
+    std::string getStackPointerAddress() const;
     std::string getIndexName(const std::string& entry) const;
     std::string getCalleeIdName(const std::string& callee) const;
     std::string getCalleeMacName(const std::string& callee) const;
diff --git a/src/sancus_support/sm_support.h b/src/sancus_support/sm_support.h
index adc798b..e5b9988 100644
--- a/src/sancus_support/sm_support.h
+++ b/src/sancus_support/sm_support.h
@@ -283,18 +283,11 @@ always_inline sm_id sancus_get_self_id(void)
  *
  * The calling module is defined as the previously executing module. That is,
  * the module that entered the currently executing module.
+ *
+ * @note This function is implemented as a compiler intrinsic to ensure the
+ * callerID remains stable in a preemptive environment.
  */
-always_inline sm_id sancus_get_caller_id(void)
-{
-    //TODO use __sm_sp_addr and an offset to retrieve the callerID stored on entry
-    sm_id ret;
-    asm(".word 0x1387\n\t"
-        "mov r15, %0"
-        : "=m"(ret)
-        :
-        : "r15");
-    return ret;
-}
+sm_id sancus_get_caller_id(void);
 
 /**
  * Perform a call to a Sancus module.
```

Reply:
Update re caller_id behavior changes:

- [x] added tests in sancus-examples to make sure this behavior will work as expected (see commit referenced above from sancus-examples repo). This will fail after merging and some assertions will have to be updated to the new behavior.
- [ ] we need to also provide a version of sancus_get_caller_id when called by unprotected code? This may be simply the instruction itself, or `unprotected_entry.S` could be updated to store the caller_id somewhere? Not sure how to select the correct function at link time -- maybe the compiler intrinsic would be easier in this respect?
- [ ] backwards compatibility will be partially broken with the new behavior; this is fine, but we need to update the exact definition of this function in the sancus_support doc above. You can prob start from your description above:
> caller_id field above the SSA. caller_id is filled after the check for an IRQ call and is only filled if it is empty (on the first call). Only when returning, caller_id is emptied. This means that during any intermediate call into the SM, the caller_id remains static until it returns.

Reply:
also, not sure, but why don't we make life easier and *do* retain backwards compatibility with a non-preemptive environment by simply overwriting the ssa_caller_id field on every SM entry or re-entry from an ocall (ie only when the call is valid and will be accepted of course)?

This would be compatible with the current definition and would not break any existing applications afais?

Arguments against this proposed behavior?

Reply:
I'm not completely sure what you mean with the last point, to overwrite the ssa_caller_id on every call. That is basically already what happens. See https://github.com/sancus-tee/sancus-compiler/pull/33/files#diff-b4f90c1290867b2f6af0d36a10035e6735c84fb624c69d6230f75aa081dab112R67 Or do you want to overwrite the HW caller ID field based on the SSA entry? That would not be the simplest solution I'd say.

My argument about breaking backwards compatibility was that we will need to include the name of the SM into the function call (either into the name or as an argument). And that would break bw compatibility. I do not think that using a compiler intrinsic is the best approach as it would completely hide away the actual function that is used from the developer. I think adding an inline `sm_##name##_get_original_caller_id` function when defining the SM would be a simple and effective solution that also forces the developer to think about what they're calling.

I can still go for the compiler intrinsic version similar to your above code if you prefer. Should be an easy commit since we don't need the additional calculations since we have the sm_##name##_caller_id address already available.

Reply:
no, we don't want to have the developer using another function. Hiding this from the developer in a compiler intrinsic is a desirable solution I'd say :)

what I meant w backwards compatibility is purely about the semantics of the returned ID: I am a bit confused w your description above:

> This means that during any intermediate call into the SM, the caller_id remains static until it returns.

So, the question is: is the returned ID (1) the one of the previously executing SM in this thread (ie as would be when no scheduler and IRQs would be involved), or (2) is the ID the one of the *original* caller of this ECALL and not updated on OCALLs as I was under the impression from your desciption? My point is we want (1) not (2), but not sure what we have atm?

To make it specific: will the tests I added in sancus-examples continue to work or will they fail (as I thought when adding the TODO lines)?

https://github.com/sancus-tee/sancus-examples/blob/master/arithmetic/foo.c#L19

Reply:
from a quick look, it seems like it's behavior (1) above, so that's good?

but the entry stub is currently very hard to read w many flows and Im not sure Aion covers all of them atm? Let's try to clarify at least in comments the atomic periods: how long is the HW entry atomic period, from where to where is it extended with clix, how does __sm_isr handling fit in the clix period etc?

Re the latter, can/should we make a stub without ISR handlign for Aion?

Reply:
also,re changes to sm_entry/exit, linking 2 outstanding severe bug issues that ideally should be fixed here as well: #13 and #22

Reply:
especially atomicity requirements for linking 2 SMs with sancus_verify as outlined in #22 are relevant to the Aion design and not yet properly handled in the current implementation afaik :thinking:

Reply:
after discussion with @fritzalder : local attestation and jumping to another SM entry point only needs to happen atomically *if* the callee may unload itself using `sancus_disable` *without* notifying potential callers first.

We concluded this is not a concern to be transparently fixed by hardware/runtime at the moment, i.e., secure linking is currently *not* meant to be interrupt-safe. This PR should add a short note for the application developer in the documentation of `sancus_disable` to clarify that SMs should not call this function when they might still be called by a trusted caller SM (and the caller SM can verify this is indeed not the case by local attestation of the callee SM code section).

Reply:
I added a warning to sancus_disable and some comments to the entry stub. Feel free to do another iteration over it.

Reply:
done, clarified a bit further sancus_disable comment

Reply:
compiler intrinsic for `sancus_get_caller_id` looks good! 

I realize we only need a solution still for `sancus_verify_caller`. However, we should drop this primitive as it can easily be implemented with other primitives as follows: `sancus_verify(adrs) + sancus_get_id(adrs) == sancus_get_caller_id()`

Reply:
disabled in sancus_support.h and left a note as per 33a1d1a 

In principle we can transparently handle this case with support from the sm_entry enclave runtime, but I don't think anyone actually uses sancus_verify_caller(). If so, we can always do this later.